sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java

```diff
@@ -0,0 +1,117 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.kafka;
+
+import java.util.List;
+import org.apache.beam.sdk.extensions.protobuf.ProtoMessageSchema;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+public class BeamKafkaProtoTable extends BeamKafkaTable {
+  private final Class<?> protoClass;
+
+  public BeamKafkaProtoTable(
+      Schema messageSchema, String bootstrapServers, List<String> topics, Class<?> protoClass) {
+    super(inferAndVerifySchema(protoClass, messageSchema), bootstrapServers, topics);
+    this.protoClass = protoClass;
+  }
+
+  @Override
+  public PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>> getPTransformForInput() {
+    return new ProtoRecorderDecoder(schema, protoClass);
+  }
+
+  @Override
+  public PTransform<PCollection<Row>, PCollection<KV<byte[], byte[]>>> getPTransformForOutput() {
+    return new ProtoRecorderEncoder(protoClass);
+  }
+
+  private static Schema inferAndVerifySchema(Class<?> protoClass, Schema messageSchema) {
+    Schema inferredSchema = new ProtoMessageSchema().schemaFor(TypeDescriptor.of(protoClass));
+    if (!messageSchema.equivalent(inferredSchema)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Given message schema '%s' does not match schema inferred from protobuf class. Protobuf class: '%s' Inferred schema: '%s'",
+              messageSchema, protoClass.getCanonicalName(), inferredSchema));
+    }
+    return inferredSchema;
+  }
+
+  /** A PTransform to convert {@code KV<byte[], byte[]>} to {@link Row}. */
+  private static class ProtoRecorderDecoder
+      extends PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>> {
+    private final Schema schema;
+    private final Class<?> clazz;
+
+    ProtoRecorderDecoder(Schema schema, Class<?> clazz) {
+      this.schema = schema;
+      this.clazz = clazz;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollection<KV<byte[], byte[]>> input) {
+      // We are not allowed to use non-vendored protobuf Message here to extend the wildcard
+      @SuppressWarnings({"unchecked", "rawtypes"})
```

Comment:
I don't think these are necessary any more are they?

Reply:
Unfortunately they are. I didn't find a solution to get rid of that. I moved the class check to ProtoMessageSchema - more below.